exes/button/src/main.cpp

```diff
@@ -0,0 +1,150 @@
+#include <cstdlib>
+#include <iostream>
+
+#include <boost/asio.hpp>
+#include <boost/program_options.hpp>
+#include <boost/sml.hpp>
+
+#include <tfc/ipc.hpp>
+#include <tfc/logger.hpp>
+#include <tfc/progbase.hpp>
+
+namespace asio = boost::asio;
+
+namespace sml = boost::sml;
+using sml::operator""_s;
+using sml::operator""_e;
+
+using std::chrono::milliseconds;
+using std::chrono_literals::operator""ms;
+using std::chrono::system_clock;
+
+// Events
+struct high {};
+struct low {};
+struct timeout {};
+
+struct my_deps {
+  my_deps(asio::io_context& ctx_para, tfc::logger::logger& logger_para, tfc::ipc_ruler::ipc_manager_client& client)
+      : ctx{ ctx_para }, logger{ logger_para }, double_signal{ ctx, client, "double_tap", "Button double tap" },
+        long_signal{ ctx, client, "long_press", "Button long press" }, touch_signal{ ctx, client, "touch", "Button touch" },
+        initial_time(system_clock::now()) {}
+  asio::io_context& ctx;
+  tfc::logger::logger& logger;
+  tfc::ipc::bool_signal double_signal;
+  tfc::ipc::bool_signal long_signal;
+  tfc::ipc::bool_signal touch_signal;
+  std::chrono::time_point<system_clock, std::chrono::nanoseconds> initial_time;
+};
+
+// guards
+constexpr auto initial_time_set = [](my_deps& deps) { deps.initial_time = system_clock::now(); };
+constexpr auto long_press = [](my_deps& deps) {
+  auto delta = system_clock::now() - deps.initial_time;
+  return delta > 1000ms && delta < 2500ms;
+};
+
+constexpr auto short_press = [](my_deps& deps) {
+  auto delta = system_clock::now() - deps.initial_time;
+  return delta < 1000ms;
+};
+
+constexpr auto too_long_press = [](my_deps& deps) {
+  auto delta = system_clock::now() - deps.initial_time;
+  return delta >= 2500ms;
+};
+
+// actions
+constexpr auto start_timeout = [](const auto&, auto& state_machine, auto& deps, const auto& subs) {
+  // Create a timer from a shared_ptr
+  auto deps_v = deps.value;
+  auto timer = std::make_shared<asio::system_timer>(deps_v.ctx);
+  auto left = 500ms;
+  timer->expires_after(left);
+  // Copy the timer into the callback to make another shared_ptr
+  // reference
+  timer->async_wait([&, timer_copy = timer](const std::error_code& err) {
+    if (err) {
+      return;
+    }
+    state_machine.process_event(timeout{}, deps, subs);
+  });
+};
+
+// Actions
+constexpr auto send_signal = [](asio::io_context& ctx, tfc::ipc::bool_signal& signal, tfc::logger::logger& logger) {
+  signal.async_send(true, [&](const std::error_code& err, size_t) {
+    if (err) {
+      logger.warn("Failed to send, {}", err.message());
+    }
+  });
+
+  // Create a timer from a shared_ptr
+  auto timer = std::make_shared<asio::system_timer>(ctx);
```

Comment:
my system instead of steady timer?

Reply:
My thought was that steady timer is not wall time. But we are interacting with the physical world ( someones finger ) but with such large timers I don't think we will find any difference.

The only guarantee steady timer gives is that every call to the timer will return a later point then the last.

Reply:
But then again if the computer applies time corrections during the time I am not sure what the behavior is.

Reply:
I think it makes sense to use steady timer for deadline timers. System timer for absolute timers, for example 24.december at 12

Reply:
I think you are correct, it looks like steady_timer is a rename of an older concept called `deadline_timer`. Will fix.